### DISCUSSION TITLE
Helper functions

### DISCUSSION THREAD

**Discussion by stevengogogo** [stevengogogo/DSA_EmailSearcher#27] [Ideas 解題思路]
`api.init`
----------

- Only can be called at beginning 30 *milliseconds*

#### Usage

```c
void api.init(
            int *n_mails,
            int *n_queries,
            mail **mails,
            query **queries
)
```

- [`api.init()`](https://github.com/stevengogogo/DSA_EmailSearcher/blob/a62565a8250a66c1ffd75ecf144cb847a4a8d058/src/api.h#L58): a struct `api` with function pointer

`api.answer`
-------------
- A prototype 

```c
void api.answer(int query_id, int answer[], int answer_length)
```

We need to output something like: [testdata/test.ans](https://cool.ntu.edu.tw/courses/4621/files/700971/download?wrap=1) (project template #13)

|Query|Answer|
||![answer](https://user-images.githubusercontent.com/29009898/119064540-2059da00-ba0e-11eb-8b8d-5cd00f1ea620.png)|

**Comment by stevengogogo** [comment]
See #24 for details about `mail` and `query`

**Comment by jcgs2503** [comment]
Pretty weird. In DSA2021_final_project (1).pdf it states that "calling the initialization function after the first 15 ms will result in Security Error." However, Data Structure and Algorithm, Spring 2021 Final Project - Email Searcher 2.0 Spec allows us to call in 30 ms.

**Reply by stevengogogo** [reply]
Timings of calling Initial function are different in this two files

|Competition & Report|Spec|
|---|---|
|![Screen Shot 2021-05-22 at 11 01 43 AM](https://user-images.githubusercontent.com/29009898/119212773-1b1f8c80-baed-11eb-8022-d0a352e975ac.png)|![Screen Shot 2021-05-22 at 11 02 57 AM](https://user-images.githubusercontent.com/29009898/119212793-4609e080-baed-11eb-8b3b-0cc61779c0db.png)|
|FIrst 15 ms|First 30 ms|

**Reply by jcgs2503** [reply]
So are they talking about different things

**Reply by stevengogogo** [reply]
I have asked on [NTU COOL](https://cool.ntu.edu.tw/courses/4621/discussion_topics/36365)

**Reply by jcgs2503** [reply]
Sure. Thanks.

**Reply by stevengogogo** [reply]
BTW, I have asked TA 塗 about the initialization:
She gave no clear situation when late initiation can happen.... 
something about preprocessing 🤷‍♂️

**Reply by stevengogogo** [reply]
In [this issue](https://github.com/stevengogogo/DSA_EmailSearcher/discussions/21#discussioncomment-770682),

There is only 15 ms mentioned in the `api.c`
